src/goto-instrument/contracts/contracts.cpp

```diff
@@ -898,20 +975,144 @@ void code_contractst::apply_loop_contract(
+    exprt invariant = static_cast<const exprt &>(
+      loop_end->get_condition().find(ID_C_spec_loop_invariant));
+    exprt decreases_clause = static_cast<const exprt &>(
+      loop_end->get_condition().find(ID_C_spec_decreases));
+
+    if(invariant.is_nil())
+    {
+      if(decreases_clause.is_not_nil() || assigns_clause.is_not_nil())
+      {
+        invariant = true_exprt{};
+        // assigns clause is missing; we will try to automatic inference
+        log.warning()
+          << "The loop at " << loop_head->source_location().as_string()
+          << " does not have an invariant in its contract.\n"
+          << "Hence, a default invariant ('true') is being used.\n"
+          << "This choice is sound, but verification may fail"
+          << " if it is be too weak to prove the desired properties."
+          << messaget::eom;
+      }
+    }
+    else
+    {
+      invariant = conjunction(invariant.operands());
+      if(decreases_clause.is_nil())
+      {
+        log.warning() << "The loop at "
+                      << loop_head->source_location().as_string()
+                      << " does not have a decreases clause in its contract.\n"
+                      << "Termination of this loop will not be verified."
+                      << messaget::eom;
+      }
+    }
+
+    const auto idx = loop_nesting_graph.add_node(
+      loop_content,
+      loop_head,
+      loop_end,
+      assigns_clause,
+      invariant,
+      decreases_clause);
+
+    if(
+      assigns_clause.is_nil() && invariant.is_nil() &&
+      decreases_clause.is_nil())
+      continue;
+
+    to_check_contracts_on_children.push_back(idx);
+
+    // If the loop has contracts to be enforced, then
+    // reject irregular loops that have "holes" within their body,
+    // while allowing instructions that are okay to be "out of loop".
+    for(auto i = loop_head; i < loop_end; ++i)
+    {
+      if(loop_content.contains(i))
+        continue;
+
+      // `break` or `goto` out of loop, are okay
+      if(i->is_goto() && !loop_content.contains(i->get_target()))
+        continue;
+
+      // `assume false`, is okay
+      if(i->is_assume() && i->get_condition().is_false())
+        continue;
+
+      // `set return value` followed by `dead`s, are okay
+      if(i->is_set_return_value())
+      {
+        do
+          i++;
+        while(i->is_dead());
+
+        // because we increment `i` in the outer `for` loop
+        i--;
+
+        // the `goto` for `return` will be handled in next iteration
+        continue;
+      }
+
+      log.error() << "Cannot enforce contracts on irregular loop at: "
+                  << loop_head->source_location()
+                  << ".\nLoop body must be a contiguous set of instructions."
+                  << messaget::eom;
+      log.debug() << "Out-of-loop-body instruction:" << messaget::eom;
+      goto_function.body.output_instruction(ns, function_name, log.debug(), *i);
+      throw 0;
+    }
```

Comment:
This is not covered by regression tests because is design to catch inconsistent GOTO programs? I'd add a comment if that's true.

Reply:
Yes exactly. I tried to, but couldn't write loops in C that would trigger this.

I will add a comment.